chat-android/src/main/java/com/ably/chat/Logger.kt

```diff
@@ -0,0 +1,119 @@
+package com.ably.chat
+
+import android.util.Log
+import java.time.LocalDateTime
+
+fun interface LogHandler {
+    fun log(message: String, level: LogLevel, throwable: Throwable?, context: LogContext)
+}
+
+data class LogContext(
+    val tag: String,
+    val clientId: String,
+    val instanceId: String,
+    val contextMap: Map<String, String> = mapOf(),
+)
```

Comment:
```suggestion
data class LogContext(
    val tag: String,
    val clientId: String,
    val instanceId: String,
    val contextMap: Map<String, String> = mapOf(),
)
```
```suggestion
data class LogContext(
    val tag: String,
    val chatClientId: String,
    val clientId: String,
    val connectionId: String?,  // null means connection is not active, retrying or failed depending on connectionState
    val connectionState: ConnectionState,  // denotes passed connection state
    val contextMap: Map<String, String> = mapOf(),
)
```

Reply:
We should have data class for `RoomContext` as well, which will be derived from `ChatClientContext`
```
val chatClientLogger = Logger(ChatClientContext());
val roomLogger = chatClientLogger.deriveWithContext(RoomContext(roomId, roomStatus, supportedRoomFeatures))
````

```
data class RoomContext {
   val roomId: String,
   val roomStatus: RoomStatus,
   val features: listOf<String>(), // Supported features of a given room with feature name
   val contextMap: Map<String, String> = mapOf(),
}
```

Underlying feature can also have one more data class `RoomFeatureContext`, this will be derived from `roomLogger` when initializing given feature. 

```
val roomFeatureLogger = roomLogger.deriveWithContext(RoomFeatureContext(roomId, channel.Name, channel.Status supportedRoomFeatures))
```
```
data class RoomFeatureContext {
   val featureName: String,
   val channeName: String,
   val channelState: AblyRealtimeChannel.ChannelState,
   val contextMap: Map<String, String> = mapOf(),
}
```

Reply:
Those will be default mandatory `Context` classes for respective classes. Those will be helpful for debugging.